### PR TITLE
Add new blog post - Optimizing Tagless Final Programs

### PR DESCRIPTION
**Pull request by LukaJCB**
My first typelevel blog post 🎉

I didn't reread yet, but would love to hear some early feedback :)

### CODE REVIEW

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+Let's say we have the following program:
+
+```scala
+def program[F[_]: Apply](F: KVStore[F]): F[List[String]] =
+    (F.get("Cats"), F.get("Dogs"), F.put("Mice", "42"), F.get("Cats"))
+      .mapN((f, s, _, t) => List(f, s, t).flatten)
+```
+
+Not a very exciting program, but it has some definite optimization potential.
+Right now, if our KVStore implementation is an asychronous one with a network boundary, our program will make 4 network requests sequentially if interpreted with the standard `Apply` instance of something like `cats.effect.IO`.
```
Comment (SystemFw) [inline posts/2017-12-27-optimizing-final-tagless.md]: `s/asychronous/asynchronous`

Comment (SystemFw) [inline posts/2017-12-27-optimizing-final-tagless.md]: s/asychronous/asynchronous

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+def optimizedProgram[F[_]: Apply](mouse: String)(F: KVStore[F]): F[List[String]] =
+  optimize(wrappedProgram)(analysisInterpreter) { case (gets, puts) =>
+    puts.toList.traverse { case (k, v) => F.put(k, v) } *> gets.toList.traverse(F.get)
+  }
+```
+
+So far so good, we've managed to write a function to generically optimize tagless final programs.
+However, one of the main advantages of tagless final is that implementation and logic should be separate concerns.
+With what we have right now, we're violating the seperation, by mixing the optimization part with the program logic part.
```
Comment (jarreds) [inline posts/2017-12-27-optimizing-final-tagless.md]: s/seperation/separation

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+By using a Tuple of `Set` and `Map` as our `Monoid`, we now get all the unique keys for our `get` and `put` operations.
+Next, we can use this information to recreate our program in an optimized way.
+
+```scala
+def optimizedProgram[F[_]: Apply](F: KVStore[F]): F[List[String]] = {
+  val (gets, puts) = program(analysisInterpreter).getConst
+
+  puts.toList.traverse { case (k, v) => F.put(k, v) } *> gets.toList.traverse(F.get)
+}
+
```
Comment (tpolecat) [inline posts/2017-12-27-optimizing-final-tagless.md]: Need a closing ``` here.

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+To get the easiest example out of the way, here's how to achieve parallelism in a tagless final program:
+
+```scala
+import cats._
+import cats.implicits._
+
+def program[M[_]: FlatMap, F[_]](K: KVStore[M])(implicit P: Parallel[M, F]) =
+  for {
+    _ <- K.put("A", a)
+    x <- (K.get("B"), K.get("C")).parMapN(f)
```
Comment (gruggiero) [inline posts/2017-12-27-optimizing-final-tagless.md]: Maybe I'm missing something, but where is `f` defined?

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 5]
```diff
@@ -0,0 +1,330 @@
+---
+layout: post
+title: Optimizing Tagless Final - Saying farewell to Free
+
+meta:
+  nav: blog
+  author: LukaJCB
```
Comment (larsrh) [inline posts/2017-12-27-optimizing-final-tagless.md]: Please add yourself to the [authors file](https://github.com/typelevel/typelevel.github.com/blob/development/_data/authors.yml).

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+
+```scala
+type Program[Alg[_[_]], F[_], A] = Alg[F] => F[A]
+```
+
+The problem of Rank-N types becomes apparent when we want to write a function where we interpret our program with two different interpreters, as we did before when interpreting into `Const`:
+
+```scala
+def optimize[Alg[_[_]], F[_]: Applicative, A, M: Monoid]
+(program: Alg[F] => F[A])
```
Comment (larsrh) [inline posts/2017-12-27-optimizing-final-tagless.md]: A bit more indentation would be useful.

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+
+Fortunately there is a workaround, albeit requiring a bit more boilerplate:
+
+```scala
+trait Program[Alg[_[_]], A] {
+  def apply[F[_]: Applicative](interpreter: Alg[F]) : F[A]
+}
+
+def optimize[Alg[_[_]], F[_]: Applicative, A, M: Monoid]
+(program: Program[Alg, A])
```
Comment (larsrh) [inline posts/2017-12-27-optimizing-final-tagless.md]: Same here, please indent.

File: posts/2017-12-27-optimizing-final-tagless.md [hunk 8]
```diff
@@ -0,0 +1,330 @@
+---
+layout: post
+title: Optimizing Tagless Final - Saying farewell to Free
```
Comment (larsrh) [inline posts/2017-12-27-optimizing-final-tagless.md]: Use either a colon or an en dash (–) in the title.

**Comment by larsrh** [comment]
@LukaJCB Are you ready to merge?

**Pull request merged**